# 項目：整理 Netflix 電影演員評分資料

## 分析目標

此資料分析的目的是，整理不同類型影視作品，如喜劇片、動作片、科幻片中，各演員出演作品的平均 IMDB 評分，從而找出各類型中的高評分作品演員

本項目的目的在於練習整理資料，以便得到可供下一步分析的資料

## 簡介

原始資料記錄了所有在 2022 年 7 月美國地區可觀看的 Netflix 電視劇及電影資料

資料包含兩個檔案：`titles.csv` 和 `credits.csv`

`titles.csv` 包含電影及電視劇相關信息，包括影視作品ID、標題、類型、描述、類型、IMDB（一個國外的在線評分網站）評分等 

`credits.csv` 包含超過 7 萬名出現在 Netflix 影視作品的導演及演員訊息，包括名字、影視作品 ID、人物名、演員類型（導演/演員）等

`titles.csv` 每列的意思如下：
- id：影視作品 ID
- title：影視作品標題
- show_type：作品類型，電視節目或電影
- description：簡短描述
- release_year：發布年份
- age_certification：適齡認證
- runtime：每集電視劇或電影的長度
- genres：類型清單
- production_countries：出品國家列表
- seasons：如果是電視劇，則是季數
- imdb_id：IMDB 的 ID
- imdb_score：IMDB 的評分
- imdb_votes：IMDB 的投票數
- tmdb_popularity：TMDB 的流行度
- tmdb_score：TMDB 的評分

`credits.csv` 每列的意思如下：
- person_ID：演員 ID
- id：參與的影視作品 ID
- name：姓名
- character_name：角色姓名
- role：演員類型，演員或導演

## 讀取資料

導入資料分析所需的函式庫，並透過 Pandas 的 `read_csv` 函數，將原始資料檔 `titles.csv` 裡的資料內容，解析為 DataFrame 並賦值給變數 `original_titles`

將原始資料檔 `credits.csv` 裡的資料內容，解析為 DataFrame 並賦值給變數 `original_credits`

In [2]:
import pandas as pd

In [3]:
original_titles = pd.read_csv("titles.csv")
original_credits = pd.read_csv("credits.csv")

In [4]:
original_titles.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

In [5]:
original_credits.head()

person_id       id             name                character   role
0       3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                      Tom  ACTOR
3       3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

## 評估並清理資料

在這一部分中，將對上一部分建立的 `original_titles` 及 `original_credits` DataFrame 所包含的資料進行評估和清理

主要從兩個方面進行：結構和內容，即整齊度和乾淨度

資料的結構性問題指不符合「每個欄位為一列，每個值為一行，每種類型的觀察單位為一個表格」這三個標準；資料的內容性問題包括存在遺失資料、重複資料、無效資料等

為了區分開經過清理的資料和原始的資料，建立新的變數 `cleaned_titles`，讓它為 `original_titles` 複製出的副本，以及建立新的變數 `cleaned_credits` ，讓它為`original_credits`複製出的變數

之後的清理步驟都會被運用在 `cleaned_titles` 和 `cleaned_credits` 上

In [6]:
cleaned_titles = original_titles.copy()
cleaned_credits = original_credits.copy()

### 資料整潔度

In [7]:
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   
5   ts22164         Monty Python's Flying Circus   SHOW   
6   tm70993                        Life of Brian  MOVIE   
7   tm14873                          Dirty Harry  MOVIE   
8  tm119281                     Bonnie and Clyde  MOVIE   
9   tm98978                      The Blue Lagoon  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   
5  A British sketch comedy series with the shows ...          1969   
6  Brian Cohen is an average young Jewish man, bu...          1979   
7  When a madman dubbed 'Scorpio' terrorizes San ...          1971   
8  In the 1930s, bored waitress Bonnie Parker fal...          1967   
9  Two small children and a ship's cook survive a...          1980   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   
5             TV-14       30                       ['comedy', 'european']   
6                 R       94                                   ['comedy']   
7                 R      102              ['thriller', 'action', 'crime']   
8                 R      110                 ['crime', 'drama', 'action']   
9                 R      104               ['romance', 'action', 'drama']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   
5               ['GB']      4.0  tt0063929         8.8     73424.0   
6               ['GB']      NaN  tt0079470         8.0    395024.0   
7               ['US']      NaN  tt0066999         7.7    155051.0   
8               ['US']      NaN  tt0061418         7.7    112048.0   
9               ['US']      NaN  tt0080453         5.8     69844.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600  
5           17.617       8.306  
6           17.770       7.800  
7           12.817       7.500  
8           15.687       7.500  
9           50.324       6.156

從資料的部分 10 行來看，`cleaned_titles` 裡的 `genres` 和 `production_countries` 的欄位中包含多個值，應當進行拆分

先提取任意一個 `genres` 欄位的值進行觀察

In [8]:
cleaned_titles['genres'][1]

"['drama', 'crime']"

雖然 `genres` 表示形式是列表，但其實際類型並非字串列表，而是字串，無法直接用 `value_counts` 統計各個值出現的次數

使用 Python 內建的 `eval` 函數，它可以把字串轉換成表達式，把表示列表的字串轉換成列表本身

In [ ]:
cleaned_titles['genres'] = cleaned_titles['genres'].apply(lambda x: eval(x))
cleaned_titles['genres'][1]

['drama', 'crime']

轉換成列表後，就能用DataFrame的 `explode` 方法，把那個列的列表值拆分成單獨的行

In [10]:
cleaned_titles = cleaned_titles.explode("genres")
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation               ['US']      1.0   
1                 R      114          drama               ['US']      NaN   
1                 R      114          crime               ['US']      NaN   
2                 R      109          drama               ['US']      NaN   
2                 R      109         action               ['US']      NaN   
2                 R      109       thriller               ['US']      NaN   
2                 R      109       european               ['US']      NaN   
3                PG       91        fantasy               ['GB']      NaN   
3                PG       91         action               ['GB']      NaN   
3                PG       91         comedy               ['GB']      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

針對 `production_countries` 列也是相同流程

每個值的 `production_countries` 值並不表示單一類型，而是一系列類型

先提取任意一個 `production_countries` 欄位的值進行觀察

In [11]:
cleaned_titles['production_countries'][1]

1    ['US']
1    ['US']
Name: production_countries, dtype: object

`production_countries` 也是相同問題，雖然表示形式是列表，但其實際類型並非字串列表，而是字串，難以進行拆分

使用 `eval` 函數進行類型轉換，並檢查轉換後確實是列表類型

In [12]:
cleaned_titles['production_countries'] = cleaned_titles['production_countries'].apply(lambda s: eval(s))
cleaned_titles['production_countries'][0]

['US']

確認型別轉換完畢後，還是用 `explode` 方法，把列表值分割成單獨的行。

In [13]:
cleaned_titles = cleaned_titles.explode('production_countries')
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation                   US      1.0   
1                 R      114          drama                   US      NaN   
1                 R      114          crime                   US      NaN   
2                 R      109          drama                   US      NaN   
2                 R      109         action                   US      NaN   
2                 R      109       thriller                   US      NaN   
2                 R      109       european                   US      NaN   
3                PG       91        fantasy                   GB      NaN   
3                PG       91         action                   GB      NaN   
3                PG       91         comedy                   GB      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

處理完 `cleaned_titles` 的結構性問題後，查看 `cleaned_credits`

In [14]:
cleaned_credits.head(10)

person_id       id             name                      character   role
0       3748  tm84618   Robert De Niro                  Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster                  Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                            Tom  ACTOR
3       3739  tm84618    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                          Betsy  ACTOR
5      32267  tm84618      Peter Boyle                         Wizard  ACTOR
6     519612  tm84618   Leonard Harris      Senator Charles Palantine  ACTOR
7      29068  tm84618   Diahnne Abbott                Concession Girl  ACTOR
8     519613  tm84618      Gino Ardito             Policeman at Rally  ACTOR
9       3308  tm84618  Martin Scorsese  Passenger Watching Silhouette  ACTOR

從頭 10 行資料來看，`cleaned_credits` 資料符合“每個欄位為一列，每個值為一行，每種類型的觀察單位為一個表格”，因此不存在結構性問題

### 資料乾淨度

接下來透過 `info`，對資料內容進行大致了解

In [15]:
cleaned_titles.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64(2), 

從輸出結果來看，`cleaned_titles` 資料共有 17818 個值，`title`、`description`、`age_certification`、`genres`、`production_countries`、`seasons `、`imdb_id`、`imdb_score`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score` 欄位都存在缺失值，將在後續進行評估與清理

此外，`release_year` 表示年份，資料類型不應為數字，應為日期，所以需要進行資料格式轉換

In [16]:
cleaned_titles["release_year"] = pd.to_datetime(cleaned_titles["release_year"], format='%Y')
cleaned_titles["release_year"]

0      1945-01-01
1      1976-01-01
1      1976-01-01
2      1972-01-01
2      1972-01-01
          ...    
5847   2021-01-01
5848   2021-01-01
5849   2021-01-01
5849   2021-01-01
5849   2021-01-01
Name: release_year, Length: 17818, dtype: datetime64[ns]

In [17]:
cleaned_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


從輸出結果來看，`cleaned_credits` 資料共有 77801 個值，其中 `character` 欄位存在缺失值，將在後續進行評估與清理

此外，`person_id` 表示演員ID，資料型別不應為數字，應為字串，所以需要進行資料格式轉換

In [18]:
cleaned_credits["person_id"] = cleaned_credits["person_id"].astype("str")
cleaned_credits["person_id"]

0           3748
1          14658
2           7064
3           3739
4          48933
          ...   
77796     736339
77797     399499
77798     373198
77799     378132
77800    1950416
Name: person_id, Length: 77801, dtype: object

#### 處理缺失資料

在`cleaned_titles`中，`title`、`description`、`age_certification`、`genres`、`production_countries`、`seasons`、`imdb_id`、`imdb_score`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`變量存在缺失值

由於影視作品的標題、描述、適齡認證、發行國家、電視劇季數、IMDB 的 ID、TMDB 的流行度、TMDB 的評分，並不影響找出各個類型中的高 IMDB 評分作品演員，所以可以保留 `title`、`description`、`age_certification`、 `production_countries`、`seasons`、`imdb_id`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score` 欄位存在空缺的值

但 `imdb_score` 和 `genres`，即 IMDB 評分和類型，跟後續要做的分析息息相關

先提取出 `imdb_score` 缺失值進行查看

In [19]:
cleaned_titles.query("imdb_score.isnull()")

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
75     tm132164                       Bill Hicks: Sane Man  MOVIE   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
...         ...                                        ...    ...   
5810  tm1225897                                 Social Man  MOVIE   
5833   ts307884                                 HQ Barbers   SHOW   
5840  tm1216735                            Sun of the Soil  MOVIE   
5844  tm1074617              Bling Empire - The Afterparty  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   

                                            description release_year  \
0     This collection includes 12 World War II-era p...   1945-01-01   
75    Sane Man was filmed before Bill recorded ‘Dang...   1989-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
...                                                 ...          ...   
5810  Two competitive social media Influencers go he...   2021-01-01   
5833  When a family run barber shop in the heart of ...   2021-01-01   
5840  In 14th-century Mali, an ambitious young royal...   2022-01-01   
5844  The stars of "Bling Empire" discuss the show's...   2021-01-01   
5848  Jamie Foxx, David Alan Grier and more from the...   2021-01-01   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
75                   R       80         comedy                   US      NaN   
145               TV-G       18  documentation                   JP     12.0   
145               TV-G       18         family                   JP     12.0   
145               TV-G       18        reality                   JP     12.0   
...                ...      ...            ...                  ...      ...   
5810               NaN       96          drama                  NaN      NaN   
5833             TV-14       24         comedy                   NG      1.0   
5840               NaN       26            NaN                  NaN      NaN   
5844               NaN       35            NaN                   US      NaN   
5848             PG-13       37            NaN                   US      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
75           NaN         NaN         NaN            3.377         7.5  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
...          ...         ...         ...              ...         ...  
5810  tt20198164         NaN         NaN              NaN         NaN  
5833         NaN         NaN         NaN            0.840         NaN  
5840         NaN         NaN         NaN            1.179         7.0  
5844         NaN         NaN         NaN              NaN         NaN  
5848         NaN         NaN         NaN            1.296        10.0  

[842 rows x 15 columns]

由於缺失分析所需的核心資料 `imdb_score`，將這些值刪除，並查看刪除後該列空缺值個數和

In [20]:
cleaned_titles = cleaned_titles.dropna(subset=["imdb_score"])
cleaned_titles["imdb_score"].isnull().sum()

0

然後提取出 `genres` 缺失值進行查看

In [21]:
cleaned_titles.query("genres.isnull()")

id                                              title   type  \
1813    ts77824  My Next Guest Needs No Introduction With David...   SHOW   
1939   ts215037                              Minecraft: Story Mode   SHOW   
2386    ts74805                   A Little Help with Carol Burnett   SHOW   
2658   ts265844                                           #ABtalks   SHOW   
4274  tm1172010                                  The Lockdown Plan  MOVIE   
4648  tm1113921                                           In Vitro  MOVIE   

                                            description release_year  \
1813  TV legend David Letterman teams up with fascin...   2018-01-01   
1939  MInecraft: Story Mode is an interactive, anima...   2018-01-01   
2386  In this unscripted series starring comedy lege...   2018-01-01   
2658  #ABtalks is a YouTube interview show hosted by...   2018-01-01   
4274                                                NaN   2020-01-01   
4648  'In Vitro' is an otherworldly rumination on me...   2019-01-01   

     age_certification  runtime genres production_countries  seasons  \
1813             TV-MA       50    NaN                   US      4.0   
1939             TV-PG       52    NaN                   US      1.0   
2386              TV-G       24    NaN                   US      1.0   
2658             TV-PG       68    NaN                  NaN      1.0   
4274               NaN       49    NaN                  NaN      NaN   
4648               NaN       27    NaN                  NaN      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1813   tt7829834         7.8      5581.0            8.217         7.6  
1939  tt10498322         5.6       347.0              NaN         NaN  
2386   tt7204366         6.3       237.0            1.621         6.2  
2658  tt12635254         9.6         7.0              NaN         NaN  
4274  tt13079112         6.5         NaN              NaN         NaN  
4648  tt10545994         7.7         NaN              NaN         NaN

由於缺失分析所需的核心資料 `genres`，將這些值刪除，並查看刪除後該列空缺值個數和

In [22]:
cleaned_titles = cleaned_titles.dropna(subset=["genres"])
cleaned_titles["genres"].isnull().sum()

0

接下來評估 `cleaned_credits` 的缺失資料，其中只有 `character` 欄位存在缺失值

角色名稱並不影響找出各個類型中的高 IMDB 評分作品演員，並且此欄位缺失也有可能因為演員類別是導演，沒有對應角色，因此可以保留 `character` 欄位存在空缺的值

#### 處理重複資料

根據資料變數的意義以及內容來看，`cleaned_titles` 裡不應該存在每個欄位都相同的值，因此查看是否存在重複值

In [23]:
cleaned_titles.duplicated().sum()

0

輸出結果為 0，表示不存在重複值

接下來看看 `cleaned_credits` 資料表是否存在重複值

In [24]:
cleaned_credits.duplicated().sum()

0

輸出結果為 0，表示不存在重複值

#### 處理不一致資料

針對 `cleaned_titles`，不一致資料可能存在於 `genres` 和 `character` 欄位中，查看是否存在多個不同值指涉同一類型，以及多個不同值指涉同一國家的情況

In [25]:
cleaned_titles['genres'].value_counts()

genres
drama            3357
comedy           2419
thriller         1446
action           1339
romance          1080
crime            1066
documentation     981
family            769
animation         732
fantasy           727
european          679
scifi             647
horror            438
history           336
music             266
reality           226
war               221
sport             188
western            53
Name: count, dtype: int64

從上方看出，`genres` 列裡並不存在不一致資料，各個數值都在指涉不同的類型

但是裡面還存在空字串表示的類型，並非有效資料，因此可以進行刪除

刪除後，查看 `cleaned_titles` 裡是否還存在 `genres` 為空字串的行

In [26]:
cleaned_titles = cleaned_titles.query('genres != ""')
cleaned_titles.query('genres == ""')

Empty DataFrame
Columns: [id, title, type, description, release_year, age_certification, runtime, genres, production_countries, seasons, imdb_id, imdb_score, imdb_votes, tmdb_popularity, tmdb_score]
Index: []

接下來，針對 `production_countries` 欄位也是一樣的流程，利用 `value_counts` 方法，得到 `production_countries` 的清單裡面各個值的出現次數

In [27]:
cleaned_titles['production_countries'].value_counts()

production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
      ... 
GT       1
CU       1
LK       1
NP       1
FO       1
Name: count, Length: 108, dtype: int64

由於 `value_counts` 執行結果中有太多值，Pandas 只會預設顯示開頭和結尾的一些值

要完整展示結果，可以把 `display.max_rows` 設定為 `None`，也就是取消展示行數上限

但目前只是呼叫 `value_counts` 時才需要看完整結果，所以可以結合 `option_context`，只更改臨時上限

In [28]:
with pd.option_context('display.max_rows', None):
    print(cleaned_titles['production_countries'].value_counts())

production_countries
US         5648
IN         1610
GB         1068
JP         1046
FR          720
KR          637
ES          637
CA          608
DE          383
CN          295
MX          264
IT          224
BR          221
AU          217
TR          195
PH          192
AR          150
ID          149
BE          148
TW          133
NG          131
PL          126
ZA          103
NL          102
HK          102
CO           94
EG           93
DK           89
TH           87
SE           81
LB           70
NO           68
AE           52
IE           49
SG           47
XX           43
IL           42
RU           41
CL           35
CH           33
PS           32
BG           31
MY           30
SA           28
AT           28
IS           28
LU           27
NZ           27
PE           26
RO           25
QA           24
CZ           22
JO           19
FI           18
HU           18
UY           15
MA           15
PT           14
KH           10
KW           10
PR            9
PK 

從上述輸出結果來看，出品國家都用兩位的國家代碼來表示，除了裡面存在一個的 `Lebanon` 值

`Lebanon` 的國家代碼是 `LB`，出現了 39 次，說明此處資料不一致

`LB` 和 `Lebanon` 都在表示同一國家，需要進行統一

把 `cleaned_titles` 裡，`production_countries` 的 `"LB"` 和 `"Lebanon"` 統一為 `LB`，並檢查替換後是否還存在 `"LB"`

In [ ]:
# 對每個值"production_countries"列的列表運用上面的函數
cleaned_titles["production_countries"] = cleaned_titles["production_countries"].replace({"Lebanon": "LB"})

# 檢查"Lebanon"是否還存在
with pd.option_context('display.max_rows', None):
    print(cleaned_titles.explode('production_countries')['production_countries'].value_counts())

production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
KR     637
ES     637
CA     608
DE     383
CN     295
MX     264
IT     224
BR     221
AU     217
TR     195
PH     192
AR     150
ID     149
BE     148
TW     133
NG     131
PL     126
ZA     103
HK     102
NL     102
CO      94
EG      93
DK      89
TH      87
SE      81
LB      71
NO      68
AE      52
IE      49
SG      47
XX      43
IL      42
RU      41
CL      35
CH      33
PS      32
BG      31
MY      30
AT      28
SA      28
IS      28
LU      27
NZ      27
PE      26
RO      25
QA      24
CZ      22
JO      19
FI      18
HU      18
UY      15
MA      15
PT      14
KW      10
KH      10
PK       9
PR       9
UA       8
MT       8
VN       8
LT       7
IR       7
CD       7
SU       7
TN       7
SN       6
AL       6
KE       6
GH       6
IQ       5
MU       5
CY       5
KN       4
GR       4
IO       4
SY       4
TZ       4
MC       4
GL       3
CM       3
AO       3
BS       3
HR       3
BD       3


另外，裡面還存在空字串表示的國家代碼，並非有效資料。但由於出品國家並非分析所需的關鍵訊息，所以可以保留出品國家為空的值

針對 `original_credits`，不一致資料可能存在於 `role` 中，將查看是否存在多個不同值指涉同一演員類型的情況

In [30]:
original_credits['role'].value_counts()

role
ACTOR       73251
DIRECTOR     4550
Name: count, dtype: int64

從上述輸出結果來看，`role` 只有兩種可能的值，`ACTOR` 或 `DIRECTOR`，不存在不一致資料

把這列的型別轉換為 `Category`，好處是比字串型別更節約記憶體空間，也能表示說值的型別有限

In [31]:
cleaned_credits["role"] = cleaned_credits["role"].astype("category")
cleaned_credits["role"]

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: category
Categories (2, object): ['ACTOR', 'DIRECTOR']

#### 處理無效或錯誤資料

可以透過 DataFrame 的 `describe` 方法，對數值統計資訊進行快速了解

In [32]:
original_titles.describe()

release_year      runtime      seasons   imdb_score    imdb_votes  \
count   5850.000000  5850.000000  2106.000000  5368.000000  5.352000e+03   
mean    2016.417094    76.888889     2.162868     6.510861  2.343938e+04   
std        6.937726    39.002509     2.689041     1.163826  9.582047e+04   
min     1945.000000     0.000000     1.000000     1.500000  5.000000e+00   
25%     2016.000000    44.000000     1.000000     5.800000  5.167500e+02   
50%     2018.000000    83.000000     1.000000     6.600000  2.233500e+03   
75%     2020.000000   104.000000     2.000000     7.300000  9.494000e+03   
max     2022.000000   240.000000    42.000000     9.600000  2.294231e+06   

       tmdb_popularity   tmdb_score  
count      5759.000000  5539.000000  
mean         22.637925     6.829175  
std          81.680263     1.170391  
min           0.009442     0.500000  
25%           2.728500     6.100000  
50%           6.821000     6.900000  
75%          16.590000     7.537500  
max        2274.044000    10.000000

從統計資料來看，`original_titles` 裡不存在脫離現實意義的數值

`original_credits` 由於不包含表示數值含義的欄位，因此無需用 `describe` 檢查

## 整理資料

In [47]:
cleaned_titles.head()

id        title   type  \
1   tm84618  Taxi Driver  MOVIE   
1   tm84618  Taxi Driver  MOVIE   
2  tm154986  Deliverance  MOVIE   
2  tm154986  Deliverance  MOVIE   
2  tm154986  Deliverance  MOVIE   

                                         description release_year  \
1  A mentally unstable Vietnam War veteran works ...   1976-01-01   
1  A mentally unstable Vietnam War veteran works ...   1976-01-01   
2  Intent on seeing the Cahulawassee River before...   1972-01-01   
2  Intent on seeing the Cahulawassee River before...   1972-01-01   
2  Intent on seeing the Cahulawassee River before...   1972-01-01   

  age_certification  runtime    genres production_countries  seasons  \
1                 R      114     drama                   US      NaN   
1                 R      114     crime                   US      NaN   
2                 R      109     drama                   US      NaN   
2                 R      109    action                   US      NaN   
2                 R      109  thriller                   US      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300

In [48]:
cleaned_credits.head()

person_id       id             name                character   role
0      3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1     14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2      7064  tm84618    Albert Brooks                      Tom  ACTOR
3      3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4     48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

對資料的整理，與分析方向緊密相關

此次資料分析目標是，整理不同類型的影視作品，比如喜劇片、動作片、科幻片等，演員出演作品的平均 IMDB 平均，從而找出各類型中的高評分作品演員

為了能同時獲得類型與演員資料，需要把 `cleaned_credits` 和 `cleaned_titles`，透過 id 作為鍵進行合併，因為兩個資料中 `id` 都是影視作品 ID 

In [49]:
credits_with_titles = pd.merge(cleaned_credits, cleaned_titles, on="id", how="inner")

合併後就能知道各個演員參與過的影視作品資訊


In [ ]:
credits_with_titles.head()

由於只需找出演員的作品，導演不在分析範圍內，因此根據 `role` 篩選出類型為 `ACTOR` 的值，供後續分析

In [52]:
actor_with_titles = credits_with_titles.query('role == "ACTOR"')

為了找出各類型中的高評分 IMDB 作品演員，需要先根據類型和演員進行分組

對演員進行分組時，選擇使用 `person_id` 而不是 `name` 欄位，因為名字可能會有重複的情況，演員 ID 會比姓名來得更精準

In [57]:
groupby_genres_and_person_id = actor_with_titles.groupby(["genres", "person_id"])

分組後，只需要對 `imdb_score` 的值進行聚合計算，因此只提出 `imdb_score` 欄位，使用 `mean` 來計算各類型作品中，每位演員的作品平均 IMDB 評分

In [61]:
score_groupby_genres_and_id = groupby_genres_and_person_id["imdb_score"].mean()
score_groupby_genres_and_id

genres   person_id
action   1000         6.866667
         100007       7.000000
         100013       6.400000
         100019       6.500000
         100020       6.500000
                        ...   
western  993735       6.500000
         998673       7.300000
         998674       7.300000
         998675       7.300000
         99940        4.000000
Name: imdb_score, Length: 168881, dtype: float64

使用 `reset_index`，對層次化索引進行重置，可以得到更規整的 DataFrame

In [62]:
score_groupby_genres_and_id_df = score_groupby_genres_and_id.reset_index()
score_groupby_genres_and_id_df

genres person_id  imdb_score
0        action      1000    6.866667
1        action    100007    7.000000
2        action    100013    6.400000
3        action    100019    6.500000
4        action    100020    6.500000
...         ...       ...         ...
168876  western    993735    6.500000
168877  western    998673    7.300000
168878  western    998674    7.300000
168879  western    998675    7.300000
168880  western     99940    4.000000

[168881 rows x 3 columns]

現在針對分類以及演員的 IMDB 評分資料已經整理好，可以進入後續的分析步驟

但當前仍然可以做一些資料整理，比如找到各類型的演員作品最高的平均分數是多少、最高分數對應的演員名字是誰

使用 `genres` 進行分組, 取出 `imdb_scrore` 欄位，計算最大值

In [64]:
genres_max_scores = score_groupby_genres_and_id_df.groupby("genres")["imdb_score"].max()
genres_max_scores

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

得到最高分後，可以跟 `imdb_score_groupby_genres_and_person_id_df` 進行合併，得到最高分對應的各個演員 ID 是甚麼，也就是這個最高平均分是哪位演員拿到的

In [65]:
genres_max_score_with_id = score_groupby_genres_and_id_df.merge(genres_max_scores, on=["genres", "imdb_score"])
genres_max_score_with_id 

genres person_id  imdb_score
0     action     12790         9.3
1     action      1303         9.3
2     action     21033         9.3
3     action    336830         9.3
4     action     86591         9.3
..       ...       ...         ...
131      war    826547         8.8
132  western     22311         8.9
133  western     28166         8.9
134  western     28180         8.9
135  western     93017         8.9

[136 rows x 3 columns]

最高分的演員不只有一位，可能有多位演員的平均分相同

為了找出演員 ID 所對應的名字，可以使用 `cleaned_credits` 進行合併

由於只需要 `person_id` 和 `name` 兩個欄位，所以選擇這兩個欄位，並把重複行刪除

In [66]:
actor_id_with_names = cleaned_credits[['person_id', 'name']].drop_duplicates()
actor_id_with_names.head(10)

person_id             name
0      3748   Robert De Niro
1     14658     Jodie Foster
2      7064    Albert Brooks
3      3739    Harvey Keitel
4     48933  Cybill Shepherd
5     32267      Peter Boyle
6    519612   Leonard Harris
7     29068   Diahnne Abbott
8    519613      Gino Ardito
9      3308  Martin Scorsese

把 `actor_id_with_names` 和 `genres_max_score_with_id` 進行合併，新增 `name` 欄位，用來顯示演員名字

In [67]:
genres_max_score_with_actor_name = genres_max_score_with_id.merge(actor_id_with_names, on="person_id")
genres_max_score_with_actor_name

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
1     action      1303         9.3       Jessie Flower
2     action     21033         9.3          Zach Tyler
3     action    336830         9.3     André Sogliuzzo
4     action     86591         9.3       Cricket Leigh
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
132  western     22311         8.9     Koichi Yamadera
133  western     28166         8.9  Megumi Hayashibara
134  western     28180         8.9      Unsho Ishizuka
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]

使用 `sort_values` 將 `genres` 進行排序，使用 `reset_index` 將索引重新排序，重新排序後 DataFrame 會多出 `index` 的欄位，使用 `drop` 刪除這個欄位

In [70]:
genres_max_score_with_actor_name = genres_max_score_with_actor_name.sort_values("genres").reset_index().drop("index", axis=1)
genres_max_score_with_actor_name

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
1     action      1303         9.3       Jessie Flower
2     action     21033         9.3          Zach Tyler
3     action    336830         9.3     André Sogliuzzo
4     action     86591         9.3       Cricket Leigh
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
132  western     28180         8.9      Unsho Ishizuka
133  western     22311         8.9     Koichi Yamadera
134  western     28166         8.9  Megumi Hayashibara
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]